shell/platform/linux/fl_binary_messenger_test.cc

```diff
@@ -384,3 +385,96 @@ TEST(FlBinaryMessengerTest, ReceiveMessage) {
+  FlBinaryMessengerResponseHandle* response_handle;
+  GMainLoop* loop;
+};
+
+static gboolean cleanup_responds_on_background_thread_info(gpointer user_data) {
+  RespondsOnBackgroundThreadInfo* info =
+      static_cast<RespondsOnBackgroundThreadInfo*>(user_data);
+  GMainLoop* loop = info->loop;
+
+  g_object_unref(info->messenger);
+  g_object_unref(info->response_handle);
+  free(info);
+
+  g_main_loop_quit(static_cast<GMainLoop*>(loop));
+
+  return G_SOURCE_REMOVE;
+}
+
+static void* response_from_thread_main(void* user_data) {
+  RespondsOnBackgroundThreadInfo* info =
+      static_cast<RespondsOnBackgroundThreadInfo*>(user_data);
+
+  fl_binary_messenger_send_response(info->messenger, info->response_handle,
+                                    nullptr, nullptr);
+
+  g_idle_add(cleanup_responds_on_background_thread_info, info);
+
+  return nullptr;
+}
+
+static void response_from_thread_cb(
+    FlBinaryMessenger* messenger,
+    const gchar* channel,
+    GBytes* message,
+    FlBinaryMessengerResponseHandle* response_handle,
+    gpointer user_data) {
+  EXPECT_NE(message, nullptr);
+  pthread_t thread;
+  RespondsOnBackgroundThreadInfo* info =
+      static_cast<RespondsOnBackgroundThreadInfo*>(
+          malloc(sizeof(RespondsOnBackgroundThreadInfo)));
+  info->messenger = FL_BINARY_MESSENGER(g_object_ref(messenger));
+  info->response_handle =
+      FL_BINARY_MESSENGER_RESPONSE_HANDLE(g_object_ref(response_handle));
+  info->loop = static_cast<GMainLoop*>(user_data);
+  EXPECT_EQ(0,
+            pthread_create(&thread, nullptr, &response_from_thread_main, info));
+}
+
+TEST(FlBinaryMessengerTest, RespondOnBackgroundThread) {
+  g_autoptr(GMainLoop) loop = g_main_loop_new(nullptr, 0);
+
+  g_autoptr(FlEngine) engine = make_mock_engine();
+  FlBinaryMessenger* messenger = fl_binary_messenger_new(engine);
+
+  // Listen for messages from the engine.
+  fl_binary_messenger_set_message_handler_on_channel(
+      messenger, "test/messages", message_cb, nullptr, nullptr);
+
+  // Listen for response from the engine.
+  fl_binary_messenger_set_message_handler_on_channel(
+      messenger, "test/responses", response_from_thread_cb, loop, nullptr);
+
+  // Trigger the engine to send a message.
+  const char* text = "Marco!";
+  g_autoptr(GBytes) message = g_bytes_new(text, strlen(text));
+  fl_binary_messenger_send_on_channel(messenger, "test/send-message", message,
+                                      nullptr, nullptr, nullptr);
+
+  // Blocks here until response_cb is called.
+  g_main_loop_run(loop);
+}
+
+TEST(FlBinaryMessengerTest, DeletingEngineClearsHandlers) {
+  FlEngine* engine = make_mock_engine();
+  g_autoptr(FlBinaryMessenger) messenger = fl_binary_messenger_new(engine);
+
+  // Listen for messages from the engine.
+  fl_binary_messenger_set_message_handler_on_channel(
+      messenger, "test/messages", message_cb, nullptr, nullptr);
+
+  ASSERT_TRUE(fl_binary_messenger_has_message_handler_on_channel(
```

Comment:
We can avoid adding this new (public) API by instead setting `destroy_notify` in the above `fl_binary_messenger_set_message_handler_on_channel` call. Then we just need to check that was called. This better reflect what we want to guarantee in the API - that the handler was notified that it is no longer connected.

Reply:
`fl_binary_messenger_has_message_handler_on_channel` is declared in the private header, it shouldn't be public, right?

Reply:
It's added a new vfunc to `shell/platform/linux/public/flutter_linux/fl_binary_messenger.h`.

Reply:
Right, that struct has an underscore as a prefix and implementing the interface isn't part of how users should be using the API.  I'm not quite sure why that is in the public header, do you know why?  Is this a gtk pattern?  I assumed it was conventionally private.

Reply:
It is public API, the `G_DECLARE_INTERFACE` macro above makes a typedef from `struct _FlBinaryMessengerInterface` to `FlBinaryMessengerInterface`. So all classes that implement this interface would now be required to implement this method.

Reply:
Thanks, made the change, done.